File: binary_tree_right.py
# Time Complexity: O(N)
# Space Complexity: O(N)
class Solution:
    def rightSideView(self, root: Optional[TreeNode]) -> List[int]:
        if not root:
            return []
        queue=deque()
        right=[]
        queue.append(root)
        while queue:
            a=0
            for i in range(len(queue)):
                popped=queue.popleft()
                if popped.left:
                    queue.append(popped.left)
                if popped.right:
                    queue.append(popped.right)
                a=popped.val
            right.append(a)
        return right

File: binary-tree-right.java
//Time Complexity: O(n)
//Space Complexity:o(n)
//Did this code successfully run on Leetcode :Yes
class Solution {
    private List<Integer> result;

    public List<Integer> rightSideView(TreeNode root) {
        result = new ArrayList<>();
        dfs(root, 0);
        return result;
    }

    private void dfs(TreeNode root, int level) {
        // edge
        if (root == null)
            return;
        int size = result.size();
        if (level == size) {
            result.add(root.val);
        } else {
            result.set(level, root.val);
        }

        // logic
        dfs(root.left, level + 1);
        dfs(root.right, level + 1);
    }
}